proposals/pytorch-elastic-proposal.md

```diff
@@ -0,0 +1,294 @@
+_Authors:_
+
+* [Ce Gao(@gaocegege)](https://github.com/gaocegege)
+
+_Status_
+
+- 2021-08-16 Draft v1
+
+## Abstract
+
+[TorchElastic](https://pytorch.org/docs/1.9.0/distributed.elastic.html), which was open sourced over a year ago in the pytorch/elastic github repository, is a runner and coordinator for PyTorch worker processes. it has been part of PyTorch core since 1.9.0. This proposal is to support such feature with the help of PyTorchJob. 
+
+## Background
+
+### TorchElastic Design
+
+TorchElastic provides the approach that makes distributed PyTorch fault-tolerant and elastic.
+
+To launch a fault-tolerant job, run the following on all nodes:
+
+```python
+python -m torch.distributed.run
+        --nnodes=NUM_NODES
+        --nproc_per_node=TRAINERS_PER_NODE
+        --rdzv_id=JOB_ID
+        --rdzv_backend=c10d
+        --rdzv_endpoint=HOST_NODE_ADDR
+        YOUR_TRAINING_SCRIPT.py (--arg1 ... train script args...)
+```
+
+To launch an elastic job, run the following on at least MIN_SIZE nodes and at most MAX_SIZE nodes.
+
+```python
+python -m torch.distributed.run
+        --nnodes=MIN_SIZE:MAX_SIZE
+        --nproc_per_node=TRAINERS_PER_NODE
+        --rdzv_id=JOB_ID
+        --rdzv_backend=c10d
+        --rdzv_endpoint=HOST_NODE_ADDR
+        YOUR_TRAINING_SCRIPT.py (--arg1 ... train script args...)
+```
+
+The command needs to be run on every node. The training job will start When at least min total number of nodes have joined. The command runs a local elastic agent on the node, which is used to launch and manage underlying worker processes. The agent on the node assigns `RANK`, `LOCAL_RANK` and so on for every worker.
```

Comment:
by node you mean worker pod?

Reply:
No, just a node which manages underlying worker processes.

Reply:
Ref https://pytorch.org/docs/stable/distributed.html

```
python -m torch.distributed.launch --nproc_per_node=<X> --nnodes=<Y>
```

There are Y nodes,  X workers per node. Thus the number of worker processes should be X * Y.

In Kubernetes, we will get Y pods, X processes per pod.

Reply:
oh these are nodes from Pytorch's perspective, not k8s. Maybe say Pytorch node. But I think Pod would be more clear as this document refers to the infrastructure more than the framework.

Reply:
SGTM, will update it.